cmd/admin-heal-ops.go

```diff
@@ -0,0 +1,836 @@
+	healItemBucketMetadata               = "bucketMeta"
+	healItemObject                       = "object"
+	healItemMultipartUpload              = "multipartUpload"
+)
+
+type healResultItem struct {
+	ResultIndex int64        `json:"resultId"`
+	Type        healItemType `json:"type"`
+	Bucket      string       `json:"bucket,omitempty"`
+	Object      string       `json:"object,omitempty"`
+	Detail      string       `json:"detail,omitempty"`
+	DiskInfo    struct {
+		AvailableOn []int `json:"availableOn"`
+		HealedOn    []int `json:"healedOn"`
+	} `json:"diskInfo"`
+}
+
+// healSequenceStatus - accumulated status of the heal sequence
+type healSequenceStatus struct {
+	// lock to update this structure as it is concurrently
+	// accessed
+	updateLock *sync.RWMutex
+
+	// summary and detail for failures
+	Summary       healStatusSummary `json:"Summary"`
+	FailureDetail string            `json:"Detail,omitempty"`
+
+	// settings for the heal sequence
+	HealSettings healSettings `json:"Settings"`
+
+	// slice of unacknowledged and available heal result records
+	Items []healResultItem `json:"Items"`
+
+	// accumulated statistics of the heal sequence
+	Statistics healStatistics `json:"HealStatistics"`
+}
+
+// structure to hold state of all heal sequences in server memory
+type allHealState struct {
+	sync.Mutex
+
+	// map of heal path to heal sequence
+	healSeqMap map[string]*healSequence
+}
+
+var (
+	// global server heal state
+	globalAllHealState allHealState
+)
+
+// initAllHealState - initialize healing apparatus
+func initAllHealState(isErasureMode bool) {
+	if !isErasureMode {
+		return
+	}
+
+	globalAllHealState = allHealState{
+		healSeqMap: make(map[string]*healSequence),
+	}
+}
+
+// LaunchNewHealSequence - launches a background routine that performs
+// healing according to the healSequence argument. For each heal
+// sequence, state is stored in the `globalAllHealState`, which is a
+// map of the heal path to `healSequence` which holds state about the
+// heal sequence.
+//
+// Heal results are persisted in server memory for
+// `keepHealSeqStateDuration`. This function also launches a
+// background routine to clean up heal results after the
+// aforementioned duration.
+func (ahs *allHealState) LaunchNewHealSequence(h *healSequence) error {
+	ahs.Lock()
+	defer ahs.Unlock()
+
+	// Check if new heal sequence overaps with any existing,
+	// running sequence
+	for k, hSeq := range ahs.healSeqMap {
+		if !hSeq.hasEnded() && (strings.HasPrefix(k, h.path) ||
+			strings.HasPrefix(h.path, k)) {
+
+			return errFnHealInvalidOverlappedPaths(h.path, k)
+		}
+	}
+
+	// Add heal state and start sequence
+	ahs.healSeqMap[h.path] = h
+
+	// Launch top-level background heal go-routine
+	go h.healSequenceStart()
+
+	// Launch clean-up routine to remove this heal sequence (after
+	// it ends) from the global state after
+	// keepHealSeqStateDuration has elapsed.
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Minute):
```

Comment:
Use `time.NewTicker()` for initiating events and we should also wait on `globalServiceDoneCh` otherwise there is no clean way to close this go-routine.

Reply:
Ok, we need to do this only because when `globalServiceDoneCh` triggers the server could be restarting/reloading (i.e. without the process exiting). If the server was only shutting down, it is not needed, as there is nothing to clean up. Fixed and added a new commit (that will later be squashed).